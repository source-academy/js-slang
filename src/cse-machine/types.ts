import * as es from 'estree'

import { Environment } from '../types'

export enum InstrType {
  RESET = 'Reset',
  WHILE = 'While',
  FOR = 'For',
  ASSIGNMENT = 'Assignment',
  UNARY_OP = 'UnaryOperation',
  BINARY_OP = 'BinaryOperation',
  POP = 'Pop',
  APPLICATION = 'Application',
  BRANCH = 'Branch',
  ENVIRONMENT = 'Environment',
  ARRAY_LITERAL = 'ArrayLiteral',
  ARRAY_ACCESS = 'ArrayAccess',
  ARRAY_ASSIGNMENT = 'ArrayAssignment',
  ARRAY_LENGTH = 'ArrayLength',
  MARKER = 'Marker',
  CONTINUE = 'Continue',
  CONTINUE_MARKER = 'ContinueMarker',
  BREAK = 'Break',
  BREAK_MARKER = 'BreakMarker'
}

interface BaseInstr {
  instrType: InstrType
  srcNode: es.Node
}

export interface WhileInstr extends BaseInstr {
  test: es.Expression
  body: es.Statement
}

export interface ForInstr extends BaseInstr {
  init: es.VariableDeclaration | es.Expression
  test: es.Expression
  update: es.Expression
  body: es.Statement
}

export interface AssmtInstr extends BaseInstr {
  symbol: string
  constant: boolean
  declaration: boolean
}

export interface UnOpInstr extends BaseInstr {
  symbol: es.UnaryOperator
}

export interface BinOpInstr extends BaseInstr {
  symbol: es.BinaryOperator
}

export interface AppInstr extends BaseInstr {
  numOfArgs: number
  srcNode: es.CallExpression
}

export interface BranchInstr extends BaseInstr {
  consequent: es.Expression | es.Statement
  alternate: es.Expression | es.Statement | null | undefined
}

export interface EnvInstr extends BaseInstr {
  env: Environment
}

export interface ArrLitInstr extends BaseInstr {
  arity: number
}

export type Instr =
  | BaseInstr
  | WhileInstr
  | AssmtInstr
  | AppInstr
  | BranchInstr
  | EnvInstr
  | ArrLitInstr

export type AgendaItem = es.Node | Instr

// Special class that cannot be found on the stash so is safe to be used
// as an indicator of a breakpoint from running the ECE machine
export class ECEBreak {}

// Special value that cannot be found on the stash so is safe to be used
// as an indicator of an error from running the ECE machine
export class ECError {
  constructor(public readonly error: any) {}
}
