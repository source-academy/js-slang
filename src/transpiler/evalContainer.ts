import { NativeStorage } from '../types'
import { NATIVE_STORAGE_ID, MODULE_PARAMS_ID } from '../constants'

type Evaler = (code: string, nativeStorage: NativeStorage, moduleParams: any) => any

/*
  We need to use new Function here to ensure that the parameter names do not get
  minified, as the transpiler uses NATIVE_STORAGE_ID for access
 */

export const sandboxedEval: Evaler = new Function(
  'code',
  NATIVE_STORAGE_ID,
  MODULE_PARAMS_ID,
  `
  if (${NATIVE_STORAGE_ID}.evaller === null) {
    return eval(code);
  } else {
    return ${NATIVE_STORAGE_ID}.evaller(code);
  }
`
) as Evaler
