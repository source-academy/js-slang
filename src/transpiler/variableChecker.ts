import type * as es from 'estree'

import { UndefinedVariable } from '../errors/errors'
import assert from '../utils/assert'
import { extractIdsFromPattern } from '../utils/ast/astUtils'
import {
  isDeclaration,
  isFunctionNode,
  isModuleDeclaration,
  isPattern
} from '../utils/ast/typeGuards'

function isModuleOrRegDeclaration(node: es.Node): node is es.ModuleDeclaration | es.Declaration {
  return isDeclaration(node) || isModuleDeclaration(node)
}

function checkPattern(pattern: es.Pattern, identifiers: Set<string>): void {
  extractIdsFromPattern(pattern).forEach(id => {
    if (!identifiers.has(id.name)) throw new UndefinedVariable(id.name, id)
  })
}

/**
 * Check a function node for undefined variables. The name of the function should be included in the `identifiers` set
 * passed in.
 */
function checkFunction(
  input: es.ArrowFunctionExpression | es.FunctionDeclaration | es.FunctionExpression,
  identifiers: Set<string>
): void {
  // Add the names of the parameters for each function into the set
  // of identifiers that should be checked against
  const newIdentifiers = new Set(identifiers)
  input.params.forEach(pattern =>
    extractIdsFromPattern(pattern).forEach(({ name }) => newIdentifiers.add(name))
  )

  if (input.body.type === 'BlockStatement') {
    checkForUndefinedVariables(input.body, newIdentifiers)
  } else checkExpression(input.body, newIdentifiers)
}

function checkExpression(
  node: es.Expression | es.RestElement | es.SpreadElement | es.Property,
  identifiers: Set<string>
): void {
  const checkMultiple = (items: (typeof node | null)[]) =>
    items.forEach(item => {
      if (item) checkExpression(item, identifiers)
    })

  switch (node.type) {
    case 'ArrayExpression': {
      checkMultiple(node.elements)
      break
    }
    case 'ArrowFunctionExpression':
    case 'FunctionExpression': {
      checkFunction(node, identifiers)
      break
    }
    case 'ClassExpression': {
      checkClass(node, identifiers)
      break
    }
    case 'AssignmentExpression':
    case 'BinaryExpression':
    case 'LogicalExpression': {
      checkExpression(node.right, identifiers)
      if (isPattern(node.left)) {
        checkPattern(node.left, identifiers)
      } else {
        checkExpression(node.left, identifiers)
      }
      break
    }
    case 'MemberExpression': {
      // TODO handle super
      checkExpression(node.object as es.Expression, identifiers)
      if (node.computed) checkExpression(node.property as es.Expression, identifiers)
      break
    }
    case 'CallExpression':
    case 'NewExpression': {
      // TODO handle super
      checkExpression(node.callee as es.Expression, identifiers)
      checkMultiple(node.arguments)
      break
    }
    case 'ConditionalExpression': {
      checkMultiple([node.alternate, node.consequent, node.test])
      break
    }
    case 'Identifier': {
      if (!identifiers.has(node.name)) {
        throw new UndefinedVariable(node.name, node)
      }
      break
    }
    case 'ImportExpression': {
      checkExpression(node.source, identifiers)
      break
    }
    case 'ObjectExpression': {
      checkMultiple(node.properties)
      break
    }
    case 'Property': {
      if (isPattern(node.value)) checkPattern(node.value, identifiers)
      else checkExpression(node.value, identifiers)
      break
    }
    case 'SpreadElement':
    case 'RestElement': {
      if (isPattern(node.argument)) {
        checkPattern(node.argument, identifiers)
        break
      }
      // Case falls through!
    }
    case 'AwaitExpression':
    case 'UnaryExpression':
    case 'UpdateExpression':
    case 'YieldExpression': {
      if (node.argument) checkExpression(node.argument as es.Expression, identifiers)
      break
    }
    case 'TaggedTemplateExpression': {
      checkExpression(node.tag, identifiers)
      checkExpression(node.quasi, identifiers)
      break
    }
    case 'SequenceExpression': // Comma operator
    case 'TemplateLiteral': {
      checkMultiple(node.expressions)
      break
    }
  }
}

/**
 * Check that a variable declaration is initialized with defined variables
 * Returns false if there are undefined variables, returns the set of identifiers introduced by the
 * declaration otherwise
 */
function checkVariableDeclaration(
  node: es.VariableDeclaration,
  identifiers: Set<string>
): Set<string> {
  const output = new Set<string>()
  node.declarations.forEach(({ id, init }) => {
    if (init) {
      if (isFunctionNode(init)) {
        assert(
          id.type == 'Identifier',
          'VariableDeclaration for function expressions should be Identifiers'
        )
        // Add the name of the function to the set of identifiers so that
        // recursive calls are possible
        const localIdentifiers = new Set([...identifiers, id.name])
        checkFunction(init, localIdentifiers)
      } else if (init.type === 'ClassExpression') {
        assert(
          id.type == 'Identifier',
          'VariableDeclaration for class expressions should be Identifiers'
        )
        const localIdentifiers = new Set([...identifiers, id.name])
        checkClass(init, localIdentifiers)
      } else {
        checkExpression(init, identifiers)
      }
    }
    extractIdsFromPattern(id).forEach(({ name }) => output.add(name))
  })
  return output
}

function checkClass(node: es.ClassDeclaration | es.ClassExpression, localIdentifiers: Set<string>) {
  node.body.body.forEach(item => {
    if (item.type === 'StaticBlock') {
      checkForUndefinedVariables(item, localIdentifiers)
      return
    }

    if (item.computed) {
      assert(
        item.key.type !== 'PrivateIdentifier',
        'Computed property should not have PrivateIdentifier key type'
      )
      checkExpression(item.key, localIdentifiers)
    }

    if (item.type === 'MethodDefinition') {
      checkFunction(item.value, localIdentifiers)
    } else if (item.value) {
      checkExpression(item.value, localIdentifiers)
    }
  })
}

/**
 * Check that the given declaration contains no undefined variables. Returns the set of identifiers
 * introduced by the node
 * @param node
 * @param identifiers
 * @returns
 */
function checkDeclaration(
  node: es.Declaration | es.ModuleDeclaration,
  identifiers: Set<string>
): Set<string> {
  switch (node.type) {
    case 'ClassDeclaration': {
      const localIdentifiers = new Set([...identifiers, node.id!.name])
      checkClass(node, localIdentifiers)
      return new Set([node.id!.name])
    }
    case 'FunctionDeclaration': {
      // Add the name of the function to the set of identifiers so that
      // recursive calls are possible
      const localIdentifiers = new Set([...identifiers, node.id!.name])
      checkFunction(node, localIdentifiers)
      return new Set([node.id!.name])
    }
    case 'VariableDeclaration':
      return checkVariableDeclaration(node, identifiers)
    case 'ImportDeclaration':
    case 'ExportAllDeclaration':
      return new Set()
    case 'ExportDefaultDeclaration': {
      if (isDeclaration(node.declaration)) {
        assert(
          node.declaration.type !== 'VariableDeclaration',
          'ExportDefaultDeclarations should not be associated with VariableDeclarations'
        )

        if (node.declaration.id) {
          return checkDeclaration(node.declaration, identifiers)
        }
        // TODO change declaration node type
      }
      checkExpression(node.declaration as es.Expression, identifiers)
      return new Set()
    }
    case 'ExportNamedDeclaration':
      return !node.declaration ? new Set() : checkDeclaration(node.declaration, identifiers)
  }
}

function checkStatement(
  node: Exclude<es.Statement, es.Declaration | es.BlockStatement>,
  identifiers: Set<string>
): void {
  const checkBody = (node: es.Statement, localIdentifiers: Set<string>) => {
    assert(!isDeclaration(node), `${node.type} cannot be found here!`)

    if (node.type === 'BlockStatement') checkForUndefinedVariables(node, localIdentifiers)
    else checkStatement(node, localIdentifiers)
  }

  switch (node.type) {
    case 'ExpressionStatement': {
      checkExpression(node.expression, identifiers)
      break
    }
    case 'ForStatement': {
      const localIdentifiers = new Set(identifiers)
      if (node.init) {
        if (node.init.type === 'VariableDeclaration') {
          // If the init clause declares variables, add them to the list of
          // local identifiers that the for statement should check
          const varDeclResult = checkVariableDeclaration(node.init, identifiers)
          varDeclResult.forEach(id => localIdentifiers.add(id))
        } else {
          checkExpression(node.init, localIdentifiers)
        }
      }

      if (node.test) checkExpression(node.test, localIdentifiers)
      if (node.update) checkExpression(node.update, localIdentifiers)

      checkBody(node.body, localIdentifiers)
      break
    }
    case 'ForInStatement':
    case 'ForOfStatement': {
      const localIdentifiers = new Set(identifiers)
      if (node.left.type === 'VariableDeclaration') {
        const varDeclResult = checkVariableDeclaration(node.left, identifiers)
        varDeclResult.forEach(id => localIdentifiers.add(id))
      }
      checkExpression(node.right, localIdentifiers)
      checkBody(node.body, localIdentifiers)
      break
    }
    case 'DoWhileStatement':
    case 'WhileStatement': {
      checkExpression(node.test, identifiers)
      checkBody(node.body, identifiers)
      break
    }
    case 'IfStatement': {
      checkBody(node.consequent, identifiers)
      if (node.alternate) checkBody(node.alternate, identifiers)
      checkExpression(node.test, identifiers)
      break
    }
    case 'SwitchStatement': {
      node.cases.forEach(c => {
        if (c.test) checkExpression(c.test, identifiers)
        c.consequent.forEach(stmt => checkBody(stmt, identifiers))
      })
      break
    }
    case 'LabeledStatement': {
      checkBody(node.body, identifiers)
      break
    }
    case 'ReturnStatement':
    // TODO Check why a return statement has an non expression argument
    case 'ThrowStatement': {
      checkExpression(node.argument as es.Expression, identifiers)
      break
    }
    case 'TryStatement': {
      // Check the try block
      checkForUndefinedVariables(node.block, identifiers)

      // Check the finally block
      if (node.finalizer) checkForUndefinedVariables(node.finalizer, identifiers)

      // Check the catch block
      if (node.handler) {
        const catchIds = new Set(identifiers)
        if (node.handler.param) {
          extractIdsFromPattern(node.handler.param).forEach(({ name }) => catchIds.add(name))
        }
        checkForUndefinedVariables(node.handler.body, catchIds)
      }
      break
    }
  }
}

export default function checkForUndefinedVariables(
  node: es.Program | es.BlockStatement | es.StaticBlock,
  identifiers: Set<string>
) {
  const localIdentifiers = new Set(identifiers)

  // Hoist class and function declarations
  for (const stmt of node.body) {
    switch (stmt.type) {
      case 'ClassDeclaration':
      case 'FunctionDeclaration': {
        localIdentifiers.add(stmt.id!.name)
        break
      }
      case 'ImportDeclaration': {
        stmt.specifiers.forEach(({ local: { name } }) => localIdentifiers.add(name))
        break
      }
      case 'ExportNamedDeclaration':
      case 'ExportDefaultDeclaration': {
        if (!stmt.declaration) break
        if (
          (stmt.declaration.type === 'ClassDeclaration' ||
            stmt.declaration.type === 'FunctionDeclaration') &&
          stmt.declaration.id
        ) {
          localIdentifiers.add(stmt.declaration.id.name)
        }
        break
      }
    }
  }

  for (const stmt of node.body) {
    if (isModuleOrRegDeclaration(stmt)) {
      checkDeclaration(stmt, localIdentifiers).forEach(id => localIdentifiers.add(id))
    } else if (stmt.type === 'BlockStatement') {
      checkForUndefinedVariables(stmt, localIdentifiers)
    } else {
      checkStatement(stmt, localIdentifiers)
    }
  }
}
