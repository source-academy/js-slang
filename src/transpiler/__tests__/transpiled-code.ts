import { mockContext } from '../../mocks/context'
import { parse } from '../../parser/parser'
import { Chapter } from '../../types'
import * as ast from '../../utils/ast/astCreator'
import { sanitizeAST } from '../../utils/ast/sanitizer'
import { stripIndent } from '../../utils/formatters'
import { transformImportDeclarations, transpile } from '../transpiler'

/*  DO NOT HAVE 'native[<digit>]' AS A SUBSTRING IN CODE STRINGS ANYWHERE IN THIS FILE!
 *  Some code here have a redundant '1;' as the last statement to prevent the
 *  code being tested from being transformed into eval.
 *  Check for variables being stored back by looking at all the tests.
 */
test('builtins do get prepended', () => {
  const code = '"ensure_builtins";'
  const context = mockContext(Chapter.SOURCE_4)
  const { transpiled } = transpile(parse(code, context)!, context)
  // replace native[<number>] as they may be inconsistent
  const replacedNative = transpiled.replace(/native\[\d+]/g, 'native')
  // replace the line hiding globals as they may differ between environments
  const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
  expect({ code, transpiled: replacedGlobalsLine }).toMatchSnapshot()
})

test('Ensure no name clashes', () => {
  const code = stripIndent`
    const boolOrErr = 1;
    boolOrErr[123] = 1;
    function f(callIfFuncAndRightArgs, wrap0, wrap1, wrap2,
      wrap3, wrap4, wrap5, wrap6, wrap7, wrap8, wrap9) {
      let wrap = 2;
      wrap0;wrap1;wrap2;wrap3;wrap4;wrap5;wrap6;wrap7;wrap8;wrap9;
    }
    const native = 123;
  `
  const context = mockContext(Chapter.SOURCE_4)
  const { transpiled } = transpile(parse(code, context)!, context)
  const replacedNative = transpiled.replace(/native0\[\d+]/g, 'native')
  const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
  expect(replacedGlobalsLine).toMatchSnapshot()
})

describe('test transformImportDeclarations', () => {
  type TestCase = [actualCode: string, expectedCode: string]

  const testCases: TestCase[] = [
    ['import { hi } from "rune";', 'const hi = modules.rune.hi;'],
    ['import { a, b } from "rune";', 'const a = modules.rune.a;\nconst b = modules.rune.b;'],
    ['import a from "rune";', 'const a = modules.rune.default;'],
    ['import a, { b } from "rune";', 'const a = modules.rune.default;\nconst b = modules.rune.b;'],
    ['import * as a from "rune";', 'const a = modules.rune;'],
    [
      'import { a as x, b as y } from "rune";',
      'const x = modules.rune.a;\nconst y = modules.rune.b;'
    ]
  ]

  test.each(testCases)('', (actual, expected) => {
    const expectedContext = mockContext(Chapter.LIBRARY_PARSER)
    const expectedProgram = parse(expected, expectedContext)
    if (!expectedProgram || expectedContext.errors.length > 0) {
      throw new Error('Expected program should not have parse errors')
    }

    const actualContext = mockContext(Chapter.LIBRARY_PARSER)
    const actualProgram = parse(actual, actualContext)

    if (!actualProgram || actualContext.errors.length > 0) {
      throw new Error('Expected program should not have parse errors')
    }

    const [declNodes, otherNodes] = transformImportDeclarations(
      actualProgram,
      ast.identifier('modules')
    )
    expect(otherNodes.length).toEqual(0)

    const finalProgram = ast.program(declNodes)
    expect(sanitizeAST(finalProgram)).toMatchObject(sanitizeAST(expectedProgram))
  })
})
