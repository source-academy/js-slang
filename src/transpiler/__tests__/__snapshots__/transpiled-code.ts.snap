// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensure no name clashes 1`] = `
"const native0 = $$NATIVE_STORAGE;
const callIfFuncAndRightArgs0 = native.operators.get(\\"callIfFuncAndRightArgs\\");
const boolOrErr0 = native.operators.get(\\"boolOrErr\\");
const wrap90 = native.operators.get(\\"wrap\\");
const unaryOp = native.operators.get(\\"unaryOp\\");
const binaryOp = native.operators.get(\\"binaryOp\\");
const throwIfTimeout = native.operators.get(\\"throwIfTimeout\\");
const setProp = native.operators.get(\\"setProp\\");
const getProp = native.operators.get(\\"getProp\\");
let lastStatementResult = undefined;
const globals = $$NATIVE_STORAGE[1].globals;
(( <globals redacted> ) => {
  return (() => {
    {
      const runtime = globals.previousScope.variables.get(\\"runtime\\").getValue();
      const display = globals.previousScope.variables.get(\\"display\\").getValue();
      const raw_display = globals.previousScope.variables.get(\\"raw_display\\").getValue();
      const stringify = globals.previousScope.variables.get(\\"stringify\\").getValue();
      const error = globals.previousScope.variables.get(\\"error\\").getValue();
      const prompt = globals.previousScope.variables.get(\\"prompt\\").getValue();
      const is_number = globals.previousScope.variables.get(\\"is_number\\").getValue();
      const is_string = globals.previousScope.variables.get(\\"is_string\\").getValue();
      const is_function = globals.previousScope.variables.get(\\"is_function\\").getValue();
      const is_boolean = globals.previousScope.variables.get(\\"is_boolean\\").getValue();
      const is_undefined = globals.previousScope.variables.get(\\"is_undefined\\").getValue();
      const parse_int = globals.previousScope.variables.get(\\"parse_int\\").getValue();
      const undefined = globals.previousScope.variables.get(\\"undefined\\").getValue();
      const NaN = globals.previousScope.variables.get(\\"NaN\\").getValue();
      const Infinity = globals.previousScope.variables.get(\\"Infinity\\").getValue();
      const math_abs = globals.previousScope.variables.get(\\"math_abs\\").getValue();
      const math_acos = globals.previousScope.variables.get(\\"math_acos\\").getValue();
      const math_acosh = globals.previousScope.variables.get(\\"math_acosh\\").getValue();
      const math_asin = globals.previousScope.variables.get(\\"math_asin\\").getValue();
      const math_asinh = globals.previousScope.variables.get(\\"math_asinh\\").getValue();
      const math_atan = globals.previousScope.variables.get(\\"math_atan\\").getValue();
      const math_atanh = globals.previousScope.variables.get(\\"math_atanh\\").getValue();
      const math_atan2 = globals.previousScope.variables.get(\\"math_atan2\\").getValue();
      const math_ceil = globals.previousScope.variables.get(\\"math_ceil\\").getValue();
      const math_cbrt = globals.previousScope.variables.get(\\"math_cbrt\\").getValue();
      const math_expm1 = globals.previousScope.variables.get(\\"math_expm1\\").getValue();
      const math_clz32 = globals.previousScope.variables.get(\\"math_clz32\\").getValue();
      const math_cos = globals.previousScope.variables.get(\\"math_cos\\").getValue();
      const math_cosh = globals.previousScope.variables.get(\\"math_cosh\\").getValue();
      const math_exp = globals.previousScope.variables.get(\\"math_exp\\").getValue();
      const math_floor = globals.previousScope.variables.get(\\"math_floor\\").getValue();
      const math_fround = globals.previousScope.variables.get(\\"math_fround\\").getValue();
      const math_hypot = globals.previousScope.variables.get(\\"math_hypot\\").getValue();
      const math_imul = globals.previousScope.variables.get(\\"math_imul\\").getValue();
      const math_log = globals.previousScope.variables.get(\\"math_log\\").getValue();
      const math_log1p = globals.previousScope.variables.get(\\"math_log1p\\").getValue();
      const math_log2 = globals.previousScope.variables.get(\\"math_log2\\").getValue();
      const math_log10 = globals.previousScope.variables.get(\\"math_log10\\").getValue();
      const math_max = globals.previousScope.variables.get(\\"math_max\\").getValue();
      const math_min = globals.previousScope.variables.get(\\"math_min\\").getValue();
      const math_pow = globals.previousScope.variables.get(\\"math_pow\\").getValue();
      const math_random = globals.previousScope.variables.get(\\"math_random\\").getValue();
      const math_round = globals.previousScope.variables.get(\\"math_round\\").getValue();
      const math_sign = globals.previousScope.variables.get(\\"math_sign\\").getValue();
      const math_sin = globals.previousScope.variables.get(\\"math_sin\\").getValue();
      const math_sinh = globals.previousScope.variables.get(\\"math_sinh\\").getValue();
      const math_sqrt = globals.previousScope.variables.get(\\"math_sqrt\\").getValue();
      const math_tan = globals.previousScope.variables.get(\\"math_tan\\").getValue();
      const math_tanh = globals.previousScope.variables.get(\\"math_tanh\\").getValue();
      const math_trunc = globals.previousScope.variables.get(\\"math_trunc\\").getValue();
      const math_E = globals.previousScope.variables.get(\\"math_E\\").getValue();
      const math_LN10 = globals.previousScope.variables.get(\\"math_LN10\\").getValue();
      const math_LN2 = globals.previousScope.variables.get(\\"math_LN2\\").getValue();
      const math_LOG10E = globals.previousScope.variables.get(\\"math_LOG10E\\").getValue();
      const math_LOG2E = globals.previousScope.variables.get(\\"math_LOG2E\\").getValue();
      const math_PI = globals.previousScope.variables.get(\\"math_PI\\").getValue();
      const math_SQRT1_2 = globals.previousScope.variables.get(\\"math_SQRT1_2\\").getValue();
      const math_SQRT2 = globals.previousScope.variables.get(\\"math_SQRT2\\").getValue();
      const pair = globals.previousScope.variables.get(\\"pair\\").getValue();
      const is_pair = globals.previousScope.variables.get(\\"is_pair\\").getValue();
      const head = globals.previousScope.variables.get(\\"head\\").getValue();
      const tail = globals.previousScope.variables.get(\\"tail\\").getValue();
      const is_null = globals.previousScope.variables.get(\\"is_null\\").getValue();
      const list = globals.previousScope.variables.get(\\"list\\").getValue();
      const draw_data = globals.previousScope.variables.get(\\"draw_data\\").getValue();
      const set_head = globals.previousScope.variables.get(\\"set_head\\").getValue();
      const set_tail = globals.previousScope.variables.get(\\"set_tail\\").getValue();
      const array_length = globals.previousScope.variables.get(\\"array_length\\").getValue();
      const is_array = globals.previousScope.variables.get(\\"is_array\\").getValue();
      const stream_tail = globals.previousScope.variables.get(\\"stream_tail\\").getValue();
      const stream = globals.previousScope.variables.get(\\"stream\\").getValue();
      const list_to_stream = globals.previousScope.variables.get(\\"list_to_stream\\").getValue();
      const parse = globals.previousScope.variables.get(\\"parse\\").getValue();
      const apply_in_underlying_javascript = globals.previousScope.variables.get(\\"apply_in_underlying_javascript\\").getValue();
      {
        const boolOrErr = 1;
        setProp(boolOrErr, 123, 1, 2, 0);
        const f = wrap90((callIfFuncAndRightArgs, wrap0, wrap1, wrap2, wrap3, wrap4, wrap5, wrap6, wrap7, wrap8, wrap9) => {
          let wrap = 2;
          wrap0;
          wrap1;
          wrap2;
          wrap3;
          wrap4;
          wrap5;
          wrap6;
          wrap7;
          wrap8;
          wrap9;
        }, \\"function f(callIfFuncAndRightArgs, wrap0, wrap1, wrap2, wrap3, wrap4, wrap5, wrap6, wrap7, wrap8, wrap9) {\\\\n  let wrap = 2;\\\\n  wrap0;\\\\n  wrap1;\\\\n  wrap2;\\\\n  wrap3;\\\\n  wrap4;\\\\n  wrap5;\\\\n  wrap6;\\\\n  wrap7;\\\\n  wrap8;\\\\n  wrap9;\\\\n}\\");
        const native = 123;
        globals.variables.set(\\"boolOrErr\\", {
          kind: \\"const\\",
          getValue: () => {
            return boolOrErr;
          }
        });
        globals.variables.set(\\"f\\", {
          kind: \\"const\\",
          getValue: () => {
            return f;
          }
        });
        globals.variables.set(\\"native\\", {
          kind: \\"const\\",
          getValue: () => {
            return native;
          }
        });
      }
    }
    return lastStatementResult;
  })();
})();
"
`;

exports[`builtins do get prepended 1`] = `
Object {
  "code": "\\"ensure_builtins\\";",
  "transpiled": "const native = $$NATIVE_STORAGE;
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFuncAndRightArgs\\");
const boolOrErr = native.operators.get(\\"boolOrErr\\");
const wrap = native.operators.get(\\"wrap\\");
const unaryOp = native.operators.get(\\"unaryOp\\");
const binaryOp = native.operators.get(\\"binaryOp\\");
const throwIfTimeout = native.operators.get(\\"throwIfTimeout\\");
const setProp = native.operators.get(\\"setProp\\");
const getProp = native.operators.get(\\"getProp\\");
let lastStatementResult = undefined;
const globals = $$NATIVE_STORAGE[0].globals;
(( <globals redacted> ) => {
  return (() => {
    {
      const runtime = globals.previousScope.variables.get(\\"runtime\\").getValue();
      const display = globals.previousScope.variables.get(\\"display\\").getValue();
      const raw_display = globals.previousScope.variables.get(\\"raw_display\\").getValue();
      const stringify = globals.previousScope.variables.get(\\"stringify\\").getValue();
      const error = globals.previousScope.variables.get(\\"error\\").getValue();
      const prompt = globals.previousScope.variables.get(\\"prompt\\").getValue();
      const is_number = globals.previousScope.variables.get(\\"is_number\\").getValue();
      const is_string = globals.previousScope.variables.get(\\"is_string\\").getValue();
      const is_function = globals.previousScope.variables.get(\\"is_function\\").getValue();
      const is_boolean = globals.previousScope.variables.get(\\"is_boolean\\").getValue();
      const is_undefined = globals.previousScope.variables.get(\\"is_undefined\\").getValue();
      const parse_int = globals.previousScope.variables.get(\\"parse_int\\").getValue();
      const undefined = globals.previousScope.variables.get(\\"undefined\\").getValue();
      const NaN = globals.previousScope.variables.get(\\"NaN\\").getValue();
      const Infinity = globals.previousScope.variables.get(\\"Infinity\\").getValue();
      const math_abs = globals.previousScope.variables.get(\\"math_abs\\").getValue();
      const math_acos = globals.previousScope.variables.get(\\"math_acos\\").getValue();
      const math_acosh = globals.previousScope.variables.get(\\"math_acosh\\").getValue();
      const math_asin = globals.previousScope.variables.get(\\"math_asin\\").getValue();
      const math_asinh = globals.previousScope.variables.get(\\"math_asinh\\").getValue();
      const math_atan = globals.previousScope.variables.get(\\"math_atan\\").getValue();
      const math_atanh = globals.previousScope.variables.get(\\"math_atanh\\").getValue();
      const math_atan2 = globals.previousScope.variables.get(\\"math_atan2\\").getValue();
      const math_ceil = globals.previousScope.variables.get(\\"math_ceil\\").getValue();
      const math_cbrt = globals.previousScope.variables.get(\\"math_cbrt\\").getValue();
      const math_expm1 = globals.previousScope.variables.get(\\"math_expm1\\").getValue();
      const math_clz32 = globals.previousScope.variables.get(\\"math_clz32\\").getValue();
      const math_cos = globals.previousScope.variables.get(\\"math_cos\\").getValue();
      const math_cosh = globals.previousScope.variables.get(\\"math_cosh\\").getValue();
      const math_exp = globals.previousScope.variables.get(\\"math_exp\\").getValue();
      const math_floor = globals.previousScope.variables.get(\\"math_floor\\").getValue();
      const math_fround = globals.previousScope.variables.get(\\"math_fround\\").getValue();
      const math_hypot = globals.previousScope.variables.get(\\"math_hypot\\").getValue();
      const math_imul = globals.previousScope.variables.get(\\"math_imul\\").getValue();
      const math_log = globals.previousScope.variables.get(\\"math_log\\").getValue();
      const math_log1p = globals.previousScope.variables.get(\\"math_log1p\\").getValue();
      const math_log2 = globals.previousScope.variables.get(\\"math_log2\\").getValue();
      const math_log10 = globals.previousScope.variables.get(\\"math_log10\\").getValue();
      const math_max = globals.previousScope.variables.get(\\"math_max\\").getValue();
      const math_min = globals.previousScope.variables.get(\\"math_min\\").getValue();
      const math_pow = globals.previousScope.variables.get(\\"math_pow\\").getValue();
      const math_random = globals.previousScope.variables.get(\\"math_random\\").getValue();
      const math_round = globals.previousScope.variables.get(\\"math_round\\").getValue();
      const math_sign = globals.previousScope.variables.get(\\"math_sign\\").getValue();
      const math_sin = globals.previousScope.variables.get(\\"math_sin\\").getValue();
      const math_sinh = globals.previousScope.variables.get(\\"math_sinh\\").getValue();
      const math_sqrt = globals.previousScope.variables.get(\\"math_sqrt\\").getValue();
      const math_tan = globals.previousScope.variables.get(\\"math_tan\\").getValue();
      const math_tanh = globals.previousScope.variables.get(\\"math_tanh\\").getValue();
      const math_trunc = globals.previousScope.variables.get(\\"math_trunc\\").getValue();
      const math_E = globals.previousScope.variables.get(\\"math_E\\").getValue();
      const math_LN10 = globals.previousScope.variables.get(\\"math_LN10\\").getValue();
      const math_LN2 = globals.previousScope.variables.get(\\"math_LN2\\").getValue();
      const math_LOG10E = globals.previousScope.variables.get(\\"math_LOG10E\\").getValue();
      const math_LOG2E = globals.previousScope.variables.get(\\"math_LOG2E\\").getValue();
      const math_PI = globals.previousScope.variables.get(\\"math_PI\\").getValue();
      const math_SQRT1_2 = globals.previousScope.variables.get(\\"math_SQRT1_2\\").getValue();
      const math_SQRT2 = globals.previousScope.variables.get(\\"math_SQRT2\\").getValue();
      const pair = globals.previousScope.variables.get(\\"pair\\").getValue();
      const is_pair = globals.previousScope.variables.get(\\"is_pair\\").getValue();
      const head = globals.previousScope.variables.get(\\"head\\").getValue();
      const tail = globals.previousScope.variables.get(\\"tail\\").getValue();
      const is_null = globals.previousScope.variables.get(\\"is_null\\").getValue();
      const list = globals.previousScope.variables.get(\\"list\\").getValue();
      const draw_data = globals.previousScope.variables.get(\\"draw_data\\").getValue();
      const set_head = globals.previousScope.variables.get(\\"set_head\\").getValue();
      const set_tail = globals.previousScope.variables.get(\\"set_tail\\").getValue();
      const array_length = globals.previousScope.variables.get(\\"array_length\\").getValue();
      const is_array = globals.previousScope.variables.get(\\"is_array\\").getValue();
      const stream_tail = globals.previousScope.variables.get(\\"stream_tail\\").getValue();
      const stream = globals.previousScope.variables.get(\\"stream\\").getValue();
      const list_to_stream = globals.previousScope.variables.get(\\"list_to_stream\\").getValue();
      const parse = globals.previousScope.variables.get(\\"parse\\").getValue();
      const apply_in_underlying_javascript = globals.previousScope.variables.get(\\"apply_in_underlying_javascript\\").getValue();
      {
        lastStatementResult = eval(\\"\\\\\\"ensure_builtins\\\\\\";\\");
      }
    }
    return lastStatementResult;
  })();
})();
",
}
`;
